src/critics/reference_trajectory_critic.cpp

```diff
@@ -92,13 +104,65 @@ xt::xtensor<double,
           P[0] = P1(s, 0) + u * P2_P1_diff(s, 0);
           P[1] = P1(s, 1) + u * P2_P1_diff(s, 1);
         }
-        min_dist = std::min(evaluate_dist(std::move(P), t, p), min_dist);
+        auto dist = evaluate_dist(std::move(P), t, p);
+        if (dist < min_dist) {
+          min_s = s;
+          min_dist = dist;
+        }
       }
+      max_s = std::max(max_s, min_s);
       mean_dist += min_dist;
     }
-    distances(t) = mean_dist / trajectories_points_count;
+
+    cost(t) = mean_dist / trajectories_points_count;
+  }
+
+  costs += xt::pow(cost * reference_cost_weight_, reference_cost_power_);
+
+  if (enable_nearest_path_angle_critic_) {
+    cost = xt::zeros<double>({trajectories_count});
+
+    auto path_angle_point = std::min(
+      reference_segments_count - 1,
+      max_s + nearest_path_angle_offset_);
+
+    auto goal_x = xt::view(P2, path_angle_point, 0);
```

Comment:
Can P1/P2/P3 be named a little more meaningfully if they're going to be used down here?

Reply:
> We have a problem here, I think your param handler class is clever and generally well thought out, but the issue is that we need to lock with a mutex the computeControl callback for the main controller API from the dynamic parameter updates so that we don't have both read and write operations happening simultaneously when a dynamic parameter request is coming in.

I know, as i said, first i wanted to make it work while robot not moving, then add lock

Reply:
I did named them previously more meaningfully, but it leads to clumsy math expressions. 
Definition given there in [ref](https://github.com/artofnothingness/mppic/pull/36/files/09b79fc2f0d53c7eaa586ec2c72de15b45e573b5#diff-97436b42f2eead7e305fce96deeebb3cc4df39a59832ff74119bcb2ef02d4bf5L50), and at creation of aliases too

Reply:
its just difficult to read down here- I agree its clearer for the math expressions. Can they be aliased again with references down here so that they're more meaningful semantically? 

I won't die on a hill about it, its more just a readability thing but I don't care that much

Reply:
Added comments on them